### ISSUE TITLE
Java function will not run on Mac/MacM1/Linux without JAVA_HOME changes

### ISSUE THREAD

**Issue by paulyuk** [Azure/azure-functions-core-tools#3294]
If I follow the VS Code quickstart on a Mac or Mac1, the app will start but ultimately will not run at runtime.  I get gRPC errors between host and worker like this:

```bash
Failed to attach to remote debuggee VM. Reason: java.net.ConnectException: Connection refused..Core Tools Version:       4.0.4915 Commit hash: N/A  (64-bit)
Function Runtime Version: 4.14.0.19631 
Functions:         HttpExample: [GET,POST] http://localhost:7071/api/HttpExample For detailed output, run func with --verbose flag.
[2023-02-27T18:42:17.709Z] Host lock lease acquired by instance ID '000000000000000000000000731A3487'.
[2023-02-27T18:43:12.593Z] Starting worker process failed
[2023-02-27T18:43:12.593Z] Microsoft.Azure.WebJobs.Script.Grpc: The operation has timed out.
[2023-02-27T18:43:12.661Z] Failed to start a new language worker for runtime: java.
[2023-02-27T18:43:12.661Z] Microsoft.Azure.WebJobs.Script.Grpc: The operation has timed out.
```

The reason explained to me by the team, is the `worker.config.json` under the Azure Core Tools is looking at this path which either exactly has to match your java path, or needs to be overridden by a classical JAVA_HOME environment variable on Mac or Linux.  
       "defaultExecutablePath": "%JAVA_HOME%/bin/java",
 
 Since I installed Java on my Mac M1 using homebrew using Microsoft's Adopt OpenJDK, the path returned by `which java` is actually `/usr/bin/java`.  When I try `/usr` as path it still does not work (keeping in mind the tools append `/bin/java` for you so leave that out.  We believe this is actually an alias and that might be a root cause issue.

The workaround that finally works is as follows.  Create a `local.settings.json` file with the following contents and place it in the same folder next to `host.json`.

Mac and Mac M1 (homebrew installs to /Library/Java/JavaVirtualMachines/JDKVERSION/Contents/Home/bin/java)
```json
{
  "IsEncrypted": false,
  "Values": {
    "AzureWebJobsStorage": "UseDevelopmentStorage=true",
    "FUNCTIONS_WORKER_RUNTIME": "java",
    "JAVA_HOME": "/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home"
    }
}
```

Note that I have `microsoft-17.jdk` but you'll have to replace with the path and version to your JDK.

Windows WSL2 (installs to /bin/java so home path is /)
```json
{
  "IsEncrypted": false,
  "Values": {
    "AzureWebJobsStorage": "UseDevelopmentStorage=true",
    "FUNCTIONS_WORKER_RUNTIME": "java",
    "JAVA_HOME": "/"
    }
}
```

The expected fixes could be some or all of the following:
- return a helpful java not found in JAVA_HOME error
- enable core tools to understand /usr/bin/java type paths
- update docs with workarounds/guidance for Mac and Linux per above

**Comment by paulyuk** [comment]
SO thread mentions this too: https://stackoverflow.com/questions/64032219/azure-function-in-java-does-not-work-locally

**Comment by kaibocai** [comment]
The issue is not comes from the core tools or azure functions. In above case, if you set `JAVA_HOME` to `/usr`, and try run `java -version` you will see the commands will hang forever and this should be the root cause of above issue seeing when try to run a function app. 
The workaround would be set `JAVA_HOME` to the real installation of jdk, either as an envrionment variables or set it in `local.setting.json`

**Comment by paulyuk** [comment]
Yes @kaibocai  you are right i misdiagnosed.  the issue is like you say setting to usr.  

However, I think the right behavior is to first try JAVA_HOME, but then fall back to PATH.  Most JDKs have /usr/bin in the path, and therefore your /usr/bin/java will work with NO JAVA_HOME set, just like Maven.  Could we consider that instead?

**Comment by dkontyko** [comment]
Hi, I set `JAVA_HOME` to `/Library/Java/JavaVirtualMachines/openjdk.jdk/Contents/Home` in `local.settings.java`, but I still get the same error: "Failed to start Worker Channel. Process fileName: %JAVA_HOME%/bin/java". Am I missing a step beyond updating and saving that settings file, then rerunning the function? Thanks.

**Comment by paulyuk** [comment]
@MadhuraBharadwaj-MSFT

**Comment by MonishaRudraswamy** [comment]
After adding JAVA_HOME in app settings /bin/java path is appending to it still not able execute getting below error

2024-04-17T08:46:38.848Z] System.Diagnostics.Process: An error occurred trying to start process 'C:User\Program Files\Java\Jdk11\bin/bin/java' with working directory 'C:\Users\Monisha.Rudraswamy\AppData\Local\Temp\2\azure-functions15533897715002063323'. The system cannot find the file specified.

**Comment by paulyuk** [comment]
`local.settings.json` is definitely not currently supported by `mvn azure-functions:run`.  The Azure Core Tools are not in the loop to load local.settings.json.

The **workaround** that works for me is to actually set the env var for JAVA_HOME.  E.g.

```bash
JAVA_HOME=/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home mvn azure-functions:run
```

or 

```bash
export JAVA_HOME=/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home
mvn azure-functions:run
```

**Comment by brais1912** [comment]
Hi everyone,

I'm working with a MacOS M1 and I'm trying to create an Azure function using Java and Maven.

I've followed all the steps on the Microsoft Azure documentation to run an Azure function locally both through IntelliJ and CLI with Maven but I'm not able to run it. I've also followed your workarounds, that is, setting the JAVA_HOME in worker.config.json file and setting the JAVA_HOME directly with JAVA_HOME=/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home mvn azure-functions:run through the CLI and I got the same result.

I've asked a question in Stackoverflow with more details, here it is: https://stackoverflow.com/questions/79084043/java-azure-function-in-macos-m1-does-not-work-when-running-locally-with-maven-or

Summarizing, I'm seeing the following on my logs after running mvn azure-functions:run or JAVA_HOME=/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home mvn azure-functions:run:

```
Extension bundle specified, skip install extension

Azure Functions Core Tools
Core Tools Version:       4.0.6280 Commit hash: N/A +421f0144b42047aa289ce691dc6db4fc8b6143e6 (64-bit)
Function Runtime Version: 4.834.3.22875
```

As you can see, there are no errors, it simply doesn't display any Azure function despite having my Java class correctly annotated with `@FunctionName("HttpTriggerJava")` and `@HttpTrigger`.

I've been struggling with this for more than 1 day, did anyone experience the same as me?

Thanks in advance.

**Comment by iliassoto** [comment]
> `local.settings.json` is definitely not currently supported by `mvn azure-functions:run`. The Azure Core Tools are not in the loop to load local.settings.json.
> 
> The **workaround** that works for me is to actually set the env var for JAVA_HOME. E.g.
> 
> JAVA_HOME=/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home mvn azure-functions:run
> or
> 
> export JAVA_HOME=/Library/Java/JavaVirtualMachines/microsoft-17.jdk/Contents/Home
> mvn azure-functions:run

After 6 hours of trying and getting thousands of messages saying "0 functions found, 0 functions loaded", this actually worked for me. Interestigly enough, the export of JAVA_HOME doesn't beahve the same in the "normal" Powershell and the Terminal in VSCode.
